# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [ ]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue
    
    def __abs__(self):
        return abs(self.value)

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
        return DualNumber(self.value + other.value,  self.dvalue + other.dvalue)
    
    def __sub__(self, other):
        return DualNumber(self.value - other.value, self.dvalue - other.dvalue)
  
    def __truediv__(self, other):
      if abs(other.value) == 0:
          raise ZeroDivisionError
      else:
          return Dual(self.value / other.value, self.dvalue / other.value - self.value / (other.value)**2 * other.dvalue)

    def __pow__(self, other):
        return DualNumber(self.value ** other.value, self.dvalue ** other.dvalue)
        raise NotImplementedError()



In [ ]:
# Tests

DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [ ]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [ ]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    
    return DualNumber(math.cos(x.value), -math.sin(x.value)*x.dvalue) 
    raise NotImplementedError()

def tan(x):
    return DualNumber(math.tan(x.value), x.dvalue / math.cos(x.value)**2)
    raise NotImplementedError()

def exp(x):
    return DualNumber(math.exp(x.value), math.exp(x.value)*x.dvalue)
    raise NotImplementedError()

In [ ]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [ ]:
x = 0.5
dx = 1

y = 4.2
dy = 0

a = x * y
da = y * dx + x *dy

b = math.sin(x)
db = math.cos(x) * dx

z = a + b
dzdx = da + db


print (z, dzdx)

x = DualNumber(0.5, 1)
y = DualNumber(4.2, 0)

z = x * y + sin(x)


print(z)

2.579425538604203 5.077582561890373
2.579425538604203 + 5.077582561890373ε


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

z = 2.58, dzdx = 5.077

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [ ]:
x = DualNumber(0.5, 0)
y = DualNumber(4.2, 1)

z = x * y + sin(x)
dzdy = z.dvalue

print('dz/dy:', dzdy)

dz/dy: 0.5


In [ ]:
#Tests
assert dzdy
assert type(dzdy) == float
